### PR TITLE
Fix flakiness caused by drop all rows quota

### LINKED ISSUE #2272: [Flake]: Build fails because DropAllRows quota is exhausted.
The Bigtable integration test against production failed with:

```
Actual: DropAllRows(name=projects/cloud-cpp-testing-resources/instances/test-instance/tables/table-2nbfh02e36g4mp36oksj8aplnzppgaedoevtfa1by0v5) Quota exceeded for quota group 'DropRowRangeGroup' and limit 'USER-100s' of service 
```

https://source.cloud.google.com/results/invocations/8e300f35-882e-4643-aaa3-e2cd03b66316/targets/cloud-cpp%2Fgithub%2Fubuntu%2Fcontinuous/log

It might be possible to get more quota, or maybe to reduce the usage, maybe delete the rows using normal data operations as opposed to TableAdmin operations? There are not that many rows created in these tests.

```
google/cloud/bigtable/testing/table_integration_test.cc:104: Failure
Value of: table_admin_->DropAllRows(TableTestEnvironment::table_id())
Expected: is OK
Actual: DropAllRows(name=projects/cloud-cpp-testing-resources/instances/test-instance/tables/table-2nbfh02e36g4mp36oksj8aplnzppgaedoevtfa1by0v5) Quota exceeded for quota group 'DropRowRangeGroup' and limit 'USER-100s' of service 'bigtableadmin.googleapis.com' for consumer 'project_number:936212892354'. [RESOURCE_EXHAUSTED]
[  FAILED  ] MutationIntegrationTest.DeleteFromColumnEndingAtTest (38 ms)
[ RUN      ] MutationIntegrationTest.DeleteFromFamilyTest
google/cloud/bigtable/testing/table_integration_test.cc:104: Failure
Value of: table_admin_->DropAllRows(TableTestEnvironment::table_id())
Expected: is OK
Actual: DropAllRows(name=projects/cloud-cpp-testing-resources/instances/test-instance/tables/table-2nbfh02e36g4mp36oksj8aplnzppgaedoevtfa1by0v5) Quota exceeded for quota group 'DropRowRangeGroup' and limit 'USER-100s' of service 'bigtableadmin.googleapis.com' for consumer 'project_number:936212892354'. [RESOURCE_EXHAUSTED]
[  FAILED  ] MutationIntegrationTest.DeleteFromFamilyTest (21 ms)
[ RUN      ] MutationIntegrationTest.DeleteFromRowTest
[       OK ] MutationIntegrationTest.DeleteFromRowTest (396 ms)
[----------] 12 tests from MutationIntegrationTest (3751 ms total)
```

### PR DESCRIPTION
**Pull request by coryan**
Before running each integration test we cleanup the test table, removing
all rows from it. We used `DropAllRows()` for this, but there is limited
quota for this operation and the builds are failing because we exceed
that quota. Use `BulkApply()` when the number of rows is relatively
small, which is most of the time, and fallback on DropAllRows()
otherwise.

This basic plan became a bit complicated because the version of the
emulator we can use has a bug in dealing with rows
deleted via `Apply()` or `BulkApply()`:

googleapis/google-cloud-go#1240

The bug is fixed, but that version has not made it to a usable SDK yet.
The bug does not affect rows deleted with `DropAllRows()`.

So to save `DropAllRows()` quota we will use `BulkApply()` when running
in production, and to workaround the bug use `DropAllRows()` when using
the emulator.

Since we need to know whether we are using the emulator in the
`TableIntegrationTest` fixture, refactor the function to check that.

This fixes #2272.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2355)

### CODE REVIEW

File: google/cloud/bigtable/testing/table_integration_test.cc [hunk 1]
```diff
@@ -105,7 +110,43 @@ void TableIntegrationTest::SetUp() {
       TableTestEnvironment::project_id(), TableTestEnvironment::instance_id(),
       ClientOptions());
 
-  ASSERT_STATUS_OK(table_admin_->DropAllRows(TableTestEnvironment::table_id()));
+  // We cannot use `DropAllRows()` to cleanup the table because the integration
+  // tests against production sometimes use all the "'DropRowRangeGroup' and
```
Comment (devbww) [inline google/cloud/bigtable/testing/table_integration_test.cc]: s/use/consume/ might be clearer.  I, at least, had to read the sentence twice.

File: google/cloud/bigtable/testing/table_integration_test.cc [hunk 2]
```diff
@@ -105,7 +110,43 @@ void TableIntegrationTest::SetUp() {
...
+  // limit 'USER-100s'" quota. Instead we delete the rows by using a
```
Comment (devbww) [inline google/cloud/bigtable/testing/table_integration_test.cc]: I assume you meant the sentence to run into the code.  Just repeating "BulkMutation", or saying "bulk mutation", is probably clearer.

File: google/cloud/bigtable/testing/table_integration_test.cc [hunk 3]
```diff
@@ -105,7 +110,43 @@ void TableIntegrationTest::SetUp() {
...
+  BulkMutation bulk;
+  auto table = GetTable();
+  // Bigtable does not support more than 100,000 mutations in a BulkMutation,
```
Comment (devbww) [inline google/cloud/bigtable/testing/table_integration_test.cc]: s/, if/. If/

File: google/cloud/bigtable/testing/table_integration_test.cc [hunk 4]
```diff
@@ -105,7 +110,43 @@ void TableIntegrationTest::SetUp() {
...
+  // if we had that many rows then just fallback on DropAllRows(). Most tests
+  // only have a small number of rows, so this is a good strategy to save
+  // DropAllRows() quota, and should be fast in most cases.
+  std::size_t const maximum_mutations = 100000;
+
+  for (auto const& row :
+       table.ReadRows(RowRange::InfiniteRange(), Filter::PassAllFilter())) {
+    if (not row) {
```
Comment (devbww) [inline google/cloud/bigtable/testing/table_integration_test.cc]: s/not /!/

File: google/cloud/bigtable/testing/table_integration_test.cc [hunk 5]
```diff
@@ -105,7 +110,43 @@ void TableIntegrationTest::SetUp() {
...
+        SingleRowMutation(row->row_key(), bigtable::DeleteFromRow()));
+    if (bulk.size() > maximum_mutations) {
+      break;
+    }
+  }
+
+  // The version of the emulator distributed with the Cloud SDK does not handle
+  // deleted rows correctly:
+  //   https://github.com/googleapis/google-cloud-go/issues/1240
+  // this has been fixed at HEAD, but the changes have not made it to a version
```
Comment (devbww) [inline google/cloud/bigtable/testing/table_integration_test.cc]: s/this/This/

File: google/cloud/bigtable/testing/table_integration_test.cc [hunk 6]
```diff
@@ -105,7 +110,43 @@ void TableIntegrationTest::SetUp() {
...
+  // we can use yet. So just use DropAllRows() in that case.
+  if (bulk.size() > maximum_mutations || UsingCloudBigtableEmulator()) {
+    ASSERT_STATUS_OK(
+        table_admin_->DropAllRows(TableTestEnvironment::table_id()));
```
Comment (devbww) [inline google/cloud/bigtable/testing/table_integration_test.cc]: Is it worth clearing `bulk` first?

Comment (coryan) [reply google/cloud/bigtable/testing/table_integration_test.cc]: I do not believe so, is is a local variable, the destructor will do the same thing.

File: google/cloud/bigtable/testing/table_integration_test.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/bigtable/testing/table_integration_test.cc]: Fixed.

File: google/cloud/bigtable/testing/table_integration_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/bigtable/testing/table_integration_test.cc]: Ooops, good habits are hard to break :grinning:

File: google/cloud/bigtable/testing/table_integration_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/bigtable/testing/table_integration_test.cc]: Fixed.

File: google/cloud/bigtable/testing/table_integration_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/bigtable/testing/table_integration_test.cc]: I thought I had push the fix for that, sorry.

File: google/cloud/bigtable/testing/table_integration_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/bigtable/testing/table_integration_test.cc]: Fixed.

**Comment by coryan** [comment]
PTAL.

File: google/cloud/bigtable/testing/table_integration_test.cc (hunk 6 above)
Comment (devbww) [reply google/cloud/bigtable/testing/table_integration_test.cc]: The idea is to not hold on to the 100,001 mutations during the `DropAllRows()`, but if the latter only has a trivial memory footprint then I guess it doesn't matter.

**Pull request merged**